pkg/reconciliation/reconcile_racks.go

```diff
@@ -2092,16 +2093,163 @@ func (rc *ReconciliationContext) CheckConditionInitializedAndReady() result.Reco
 	return result.Continue()
 }
 
-func (rc *ReconciliationContext) cleanupAfterScaling() error {
-	var err error
+const (
+	podJobIdAnnotation     = "cassandra.datastax.com/job-id"
+	podJobStatusAnnotation = "cassandra.datastax.com/job-status"
+	podJobHandler          = "cassandra.datastax.com/job-runner"
+	jobHandlerMgmtApi      = "management-api"
+
+	podJobCompleted = "COMPLETED"
+	podJobError     = "ERROR"
+	podJobWaiting   = "WAITING"
+)
+
+var (
+	jobRunner chan int = make(chan int, 1)
+)
+
+func (rc *ReconciliationContext) cleanupAfterScaling() result.ReconcileResult {
+	// We sort to ensure we process the pods in the same order
+	sort.Slice(rc.dcPods, func(i, j int) bool {
+		rackI := rc.dcPods[i].Labels[api.RackLabel]
+		rackJ := rc.dcPods[j].Labels[api.RackLabel]
+
+		if rackI != rackJ {
+			return rackI < rackJ
+		}
+
+		return rc.dcPods[i].Name < rc.dcPods[j].Name
+	})
+
+	for idx, pod := range rc.dcPods {
+		features, err := rc.NodeMgmtClient.FeatureSet(pod)
+		if err != nil {
+			return result.Error(err)
+		}
+
+		if pod.Annotations == nil {
+			pod.Annotations = make(map[string]string)
+		}
+
+		if podJobId, found := pod.Annotations[podJobIdAnnotation]; found {
+			if features.Supports(httphelper.AsyncSSTableTasks) {
+				_, found := pod.Annotations[podJobStatusAnnotation]
+				if !found {
+					podPatch := client.MergeFrom(pod.DeepCopy())
+					// Pod is currently processing something, or has finished processing.. update the status
+					details, err := rc.NodeMgmtClient.JobDetails(pod, podJobId)
+					if err != nil {
+						rc.ReqLogger.Error(err, "Could not get JobDetails for pod", "Pod", pod)
+						return result.Error(err)
+					}
+
+					if details.Id == "" {
+						// This job was not found, pod most likely restarted. Let's retry..
+						delete(pod.Annotations, podJobIdAnnotation)
+						err = rc.Client.Patch(rc.GetContext(), pod, podPatch)
+						if err != nil {
+							return result.Error(err)
+						}
+						return result.RequeueSoon(2)
+					} else if details.Status == podJobError {
+						// Log the error, move on
+						rc.ReqLogger.Error(fmt.Errorf("cleanup failed: %s", details.Error), "Job failed to successfully complete the cleanup", "Pod", pod)
+						pod.Annotations[podJobStatusAnnotation] = podJobError
+						err = rc.Client.Patch(rc.GetContext(), pod, podPatch)
+						if err != nil {
+							return result.Error(err)
+						}
+						continue
+					} else if details.Status == podJobCompleted {
+						// Pod has finished, remove the job_id and let us move to the next pod
+						pod.Annotations[podJobStatusAnnotation] = podJobCompleted
+						err = rc.Client.Patch(rc.GetContext(), pod, podPatch)
+						if err != nil {
+							return result.Error(err)
+						}
+						continue
+					} else if details.Status == podJobWaiting {
+						// Job is still running or waiting
+						return result.RequeueSoon(1)
```

Comment:
As we know cleanup can take a considerable amount of time. Is there a reason to use such a short delay here?

Reply:
I modified it to 10 seconds now. But I'd rather not increase more, since it slows down scale up process and blocks other changes, so we'd want it somewhat responsive despite some small additional load.